### PR TITLE
Enable adding plugins from .tgz or .tar.gz

### PR DESCRIPTION
**Pull request by Mitko-Kerezov**
Install the plugin in the temp directory, copy the plugin to project if needed and set dependency.
Ping @rosen-vladimirov @teobugslayer for review

### CODE REVIEW

**Comment by rosen-vladimirov** [comment]
```
$ appbuilder plugin fetch my.tgz
```
will not work, is this expected?

**Comment by ligaz** [comment]
image

**Comment by Mitko-Kerezov** [comment]
Ping @rosen-vladimirov for a second look,
@ligaz we can definitely think about unit tests for this class but not in this PR IMO

**Comment by rosen-vladimirov** [comment]
Current message is incorrect:
```
~/Work/icenium-cli/scratch/myNativeAppWithPlugins $ appbuilder plugin add ../my-plugin/my-plugin-1.0.0.tgz 
Successfully fetched plugin my-plugin.
```

File: lib/services/nativescript-project-plugins-service.ts [hunk 1]
```diff
@@ -188,25 +191,47 @@ export class NativeScriptProjectPluginsService implements IPluginsService {
...
+		}).future<void>()();
+	}
+
+	public fetchPluginBasicInformation(pluginIdentifier: string, failMessageMethodName: string,  opts?: { isTgz: boolean }): IFuture<IBasicPluginInformation> {
+		return ((): IBasicPluginInformation => {
 			if(!pluginIdentifier) {
 				this.$errors.fail("You must specify local path, URL to a plugin repository, name or keywords of a plugin published to the NPM.");
 			}
+			let pluginIdentifierExtname = path.extname(pluginIdentifier),
+				isTgz = opts && opts.isTgz || pluginIdentifierExtname === ".tgz" || pluginIdentifierExtname === ".gz";
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: You have this check:
```TypeScript
 if(pluginIdentifierExtname === ".tgz" || pluginIdentifierExtname === ".gz")
```
Consider extracting method: isTgz

File: lib/services/nativescript-project-plugins-service.ts [hunk 2]
```diff
@@ -188,25 +191,47 @@ export class NativeScriptProjectPluginsService implements IPluginsService {
 
 	public fetch(pluginIdentifier: string): IFuture<void> {
 		return (() => {
+			let pluginBasicInfo = this.fetchPluginBasicInformation(pluginIdentifier, "fetch").wait();
+			this.$logger.printMarkdown(util.format("Successfully fetched plugin `%s`.", pluginBasicInfo.name));
+		}).future<void>()();
+	}
+
+	public fetchPluginBasicInformation(pluginIdentifier: string, failMessageMethodName: string,  opts?: { isTgz: boolean }): IFuture<IBasicPluginInformation> {
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: the method should be private

Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: the name could be improved, but I'm really out of ideas at the moment

**Comment by rosen-vladimirov** [comment]
The code fails when .tgz file does not exist:
```
~/Work/icenium-cli/scratch/myNativeAppWithPlugins $ appbuilder plugin add test.tgz
Path must be a string. Received null
```

File: lib/services/nativescript-project-plugins-service.ts [hunk 3]
```diff
@@ -188,25 +191,47 @@ export class NativeScriptProjectPluginsService implements IPluginsService {
...
+			if (pathToInstalledPlugin) {
 				let pluginsPath = path.join(this.$project.getProjectDir().wait(), "plugins");
 				// use cp instead of mv, as it would fail if pathToInstalledPlugin is mounted
 				// on a different device from the pluginsPath with error:
 				// Error: EXDEV, cross-device link not permitted
-				shelljs.cp("-Rf", [pathToInstalledPlugin], pluginsPath);
-				let pluginBasicInfo = this.installLocalPlugin(path.join(pluginsPath, path.basename(pathToInstalledPlugin)), {addPluginToPackageJson: true}).wait();
-				this.$logger.printMarkdown(util.format("Successfully fetched plugin `%s`.", pluginBasicInfo.name));
+				shelljs.cp("-Rf", [actualPlugin], pluginsPath);
+				return this.installLocalPlugin(path.join(pluginsPath, path.basename(actualPlugin)), installLocalPluginOptions).wait();
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: We've lost `{addPluginToPackageJson: true}` option
:crying_cat_face:

Comment (Mitko-Kerezov) [inline lib/services/nativescript-project-plugins-service.ts]: It is actually defined as a property of `installLocalPluginOptions` a little above

**Comment by Mitko-Kerezov** [comment]
Issues addressed, ping @rosen-vladimirov for another **thorough** review

File: lib/services/nativescript-project-plugins-service.ts [hunk 4]
```diff
@@ -188,24 +190,12 @@ export class NativeScriptProjectPluginsService implements IPluginsService {
 
 	public fetch(pluginIdentifier: string): IFuture<void> {
 		return (() => {
-			if(!pluginIdentifier) {
-				this.$errors.fail("You must specify local path, URL to a plugin repository, name or keywords of a plugin published to the NPM.");
-			}
+			let pluginLocalPath = path.resolve(pluginIdentifier),
+				pluginLocalPathExists = this.$fs.exists(pluginLocalPath).wait(),
+				supressMessage = pluginLocalPathExists && (pluginLocalPath.indexOf(this.$project.getProjectDir().wait()) === -1 || pluginLocalPath.indexOf(NativeScriptProjectPluginsService.NODE_MODULES_DIR_NAME) !== -1);
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: sup**p**ress

**Comment by rosen-vladimirov** [comment]
:+1:

**Pull request merged**